src/librustc/traits/project.rs

```diff
@@ -1079,12 +1079,10 @@ fn assemble_candidates_from_impls<'cx, 'tcx>(
                 if !is_default {
                     true
                 } else if obligation.param_env.reveal == Reveal::All {
-                    debug_assert!(!poly_trait_ref.needs_infer());
```

Comment:
To be clear, did you see this debug assertion fail?

Reply:
Yes, in a run-pass specialization test, as I assume this can happen inside a temporary inference context inside a normalization query.

Reply:
This little bit of code is making me nervous. I'm wondering if we might get into some scenario where we have an unresolved inference variable and so we opt *not* to reveal the item. I *think* this would mean we normalize to (e.g.) `<T as Foo<?X>>::Bar`. But then later the inference variable `?X` gets resolved in such a way that we *could* have revealed the value. This could lead to an inconsistency that causes some ICEs or something.

I think I would feel better if we flagged cases that involve inference variables as 'ambiguous'. I have to remember how that is done in this code, I can check if it has any impact.

Still, I could probably be persuaded to land it as is.

Reply:
I'm pretty sure not revealing *is* "ambiguous", to the extent that projection would be.
Doesn't ambiguity result in not making progress, in general?

I don't mind ICEs caused indirectly by this, considering this was even more ICE-y before, but also keep in mind this requires specialization, which is still unstable.

Reply:
Huh, looks like the assert was added by... me, in #35091, and it assumed that only monomorphic codegen would use the `Reveal::All` codepath (I was curious and dug through the history).